visualization/mayavi_vis_utils.py

```diff
@@ -0,0 +1,115 @@
+"""
+Utilities for rendering camera frustums and 3d point clouds using the Mayavi mlab library.
+
+Author: John Lambert
+"""
+
+import argparse
+from typing import List
+
+import mayavi
+import numpy as np
+from colour import Color
+from gtsam import Cal3Bundler, Pose3
+from mayavi import mlab
+
+from gtsfm.common.view_frustum import ViewFrustum
+
+
+def draw_point_cloud_mayavi(
+    fig: mayavi.core.scene.Scene, point_cloud: np.ndarray, rgb: np.ndarray, sphere_radius: float
+) -> None:
+    """Render a point cloud as a collection of spheres, using Mayavi.
+
+    Args:
+        fig: Mayavi figure object.
+        point_cloud: array of shape (N,3) representing 3d points.
+        rgb: uint8 array of shape (N,3) representing colors in RGB order, in the range [0,255].
+        sphere_radius: radius of each rendered sphere.
+    """
+    n = point_cloud.shape[0]
+    x, y, z = point_cloud.T
+    alpha = np.ones((n, 1)).astype(np.uint8) * 255  # no transparency
+    rgba = np.hstack([rgb, alpha]).astype(np.uint8)
+
+    pts = mlab.pipeline.scalar_scatter(x, y, z)  # plot the points
+    pts.add_attribute(rgba, "colors")  # assign the colors to each point
+    pts.data.point_data.set_active_scalars("colors")
+    g = mlab.pipeline.glyph(pts)
+    g.glyph.glyph.scale_factor = sphere_radius  # set scaling for all the points
+    g.glyph.scale_mode = "data_scaling_off"  # make all the points same size
+
+
+def draw_cameras_mayavi(
+    fig: mayavi.core.scene.Scene, wTi_list: List[Pose3], calibrations: List[Cal3Bundler], zcwTw: Pose3
+) -> None:
+    """Render camera frustums as collections of line segments, using Mayavi mlab.
+
+    Args:
+        fig: Mayavi mlab figure object.
+        wTi_list: list of camera poses for each image.
+        calibrations: calibration object for each camera.
+        zcwTw: transforms world points to a new world frame where the point cloud is zero-centered.
```

Comment:
why not just accept points which are zero-centered? I dont see the point of performing centering in this function. This function should just be concerned with viewing.

Reply:
so my main idea is here -- we apply a transformation to the points, so we need to apply the same transformation to the cameras.

Would it be simpler if we just transformed the `wTI_list` before plotting the cameras?

Reply:
cool, so we don't have an explicit GtsfmData instance here, but I moved the camera pose shift to the main loop in `view_scene.py`, before passing it to the Open3d or Mayavi vis utilities.